### PR TITLE
202 fix flaky tests

### LINKED ISSUE #202: Several unit tests unstable
Prompted by some repeated unit-test running to investigate https://github.com/DiamondLightSource/dodal/issues/195, we have a range of unstable unit tests.

### Acceptance criteria

- Be able to run `while python -m pytest -m "not s03" ; do :; done` for a reasonable amount of time without any failures. (That command repeatedly runs the test suite, until something fails). 
  * Currently on my machine it takes ~a few minutes on average to run the tests enough times to get a failure.

The most common offenders seem to be:
- `test_given_pin_tip_found_before_timeout_then_timeout_status_cleaned_up_and_tip_value_remains`
- `test_given_stale_parameters_goes_high_before_callbacks_then_stale_parameters_waited_on`
- `test_unsuccessful_false_roi_mode_change_results_in_callback_error` [ex 1](https://github.com/DiamondLightSource/dodal/actions/runs/6496883388/job/17646935568) [ex 2](https://github.com/DiamondLightSource/dodal/actions/runs/6496883388/job/17646875300)
- `test_when_zoom_1_then_flat_field_applied`
- `test_when_led1_not_out_it_switches_on`

If possible, document any common patterns/anti-patterns that are causing these tests to be unstable so that we can avoid them in future.

### PR DESCRIPTION
**Pull request by dperl-dls**
Fixes #202 

### Instructions to reviewer on how to test:
1. Run `while python -m pytest --random-order -m "not s03" ; do :; done` and confirm that tests do not fail for many runs - but ignore segfault at the end of tests as long as no tests failed

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://github.com/DiamondLightSource/dodal/wiki/Device-Standards)

### CODE REVIEW

File: src/dodal/devices/utils.py [hunk 1]
```diff
@@ -47,14 +47,22 @@ def run_functions_without_blocking(
     # intermediate statuses have an exception, the full_status will timeout.
     full_status = Status(obj=associated_obj, timeout=timeout)
 
-    def closing_func(old_status):
-        check_callback_error(old_status)
-        full_status.set_finished()
+    def closing_func(old_status: Status):
+        LOGGER.info(f"Status closing_func got {old_status}")
```
Comment (DominicOram) [inline src/dodal/devices/utils.py:51]: Should: This was just for testing, we should remove now

File: src/dodal/devices/utils.py [hunk 2]
```diff
@@ -63,14 +71,15 @@ def wrap_func(old_status, current_func: Callable[[], StatusBase], next_func):
             )
             raise ValueError(f"{current_func} does not return a Status")
 
+        LOGGER.info(f"Status of run func {status}")
```
Comment (DominicOram) [inline src/dodal/devices/utils.py:74]: Should: As above, remove now

File: tests/devices/unit_tests/test_utils.py [hunk 3]
```diff
@@ -35,15 +39,32 @@ def test_full_status_gives_error_if_intermediate_status_fails():
...
 
 
 def test_wrap_function_callback():
-    dummy_func = MagicMock(return_value=Status)
-    run_functions_without_blocking([lambda: get_good_status(), dummy_func])
+    dummy_func = MagicMock(return_value=Status())
+    returned_status = run_functions_without_blocking(
+        [lambda: get_good_status(), dummy_func]
+    )
+    dummy_func.assert_called_once
```
Comment (DominicOram) [inline tests/devices/unit_tests/test_utils.py:57]: Should: You're not actually calling this?

File: tests/devices/unit_tests/test_utils.py [hunk 4]
```diff
@@ -35,15 +39,32 @@ def test_full_status_gives_error_if_intermediate_status_fails():
...
+    try:
+        returned_status.wait(0.1)
+    except BaseException:
+        pass
```
Comment (DominicOram) [inline tests/devices/unit_tests/test_utils.py:61]: Should: I think slightly cleaner for dummy_func to return a good status then we don't expect any exception

File: tests/devices/unit_tests/test_utils.py [hunk 5]
```diff
@@ -35,15 +39,32 @@ def test_full_status_gives_error_if_intermediate_status_fails():
...
+
+
+def test_wrap_function_callback_errors_on_wrong_return_type():
+    dummy_func = MagicMock(return_value=3)
+    with pytest.raises(ValueError):
+        run_functions_without_blocking([lambda: get_good_status(), dummy_func])
     dummy_func.assert_called_once
```
Comment (DominicOram) [inline tests/devices/unit_tests/test_utils.py:68]: Should: This needs to be called

File: tests/devices/unit_tests/test_utils.py [hunk 6]
```diff
@@ -35,15 +39,32 @@ def test_full_status_gives_error_if_intermediate_status_fails():
...
+    try:
+        returned_status.wait(0.1)
+    except BaseException:
+        pass
+
+
+def test_wrap_function_callback_errors_on_wrong_return_type():
+    dummy_func = MagicMock(return_value=3)
+    with pytest.raises(ValueError):
+        run_functions_without_blocking([lambda: get_good_status(), dummy_func])
```
Comment (DominicOram) [inline tests/devices/unit_tests/test_utils.py:67]: Should: We should get the returned status and wait on it to make sure it's tidied

**Review by DominicOram — CHANGES_REQUESTED** [review]
Looks good, thanks. Can you merge main and then also https://github.com/DiamondLightSource/dodal/pull/221 to confirm all the tests are passing in CI, then happy to approve.

**Comment by dperl-dls** [comment]
When I run the tests a lot of times there are still occasionally threads which don't get tidied up, but I think we should fix them separately

**Review by DominicOram — APPROVED** [review]
Great, thank you!

**Pull request merged**